### DISCUSSION TITLE
Sum of LineItems

### DISCUSSION THREAD

**Discussion by Chicagoan2016** [MarimerLLC/csla#3182] [Questions] [answered]
We are migrating a non-csla application to Csla. We have line items (collection) contained in an order object.
How should we implement total amount of line items contained in a particular order object? A business rule that gets called when we save the order object and update a property 'Totalamount'?

Regards

**Comment by TheCakeMonster** [comment]
Do you need a property on the parent? I wonder if instead you can dynamically sum the totals from the child collection.

**Reply by Chicagoan2016** [reply]
You mean without a rule?

**Reply by TheCakeMonster** [reply]
@Chicagoan2016 Yes, without a rule. There's nothing wrong with creating a rule, but it doesn't really sound like you need it. Just iterate over the items in the child collection in your Create and Update methods. Iteration over a local collection is so fast it's virtually unmeasurable. Having a rule will cause the rule to run every time any child's price changes, or any child is added or removed. That's a lot of times in a normal UI.

YAGNI.

**Comment by kcabral817** [comment]
If you have to have the total calculated after a read from the persistence store, I would lean towards a business rule.

Regards

Kevin
________________________________

**Reply by Chicagoan2016** [reply]
No, it will be calculated only on 'Save'.

**Comment by kcabral817** [comment, marked as answer]
You have some options then.  You can perform the calculation in the save method, business rule, or in the view model.  However, I would still lean towards a business rule and then you can leverage that rule if the business rule changes and folks want it calculated real-time or at some other point in the process flow.

Here are two methods that I use to intercept a property change value and perform some work.  These are in my business class and you can add an if statement to check for the property that has the values and then sum them and do it real time.

protected override void PropertyHasChanged(IPropertyInfo property)
        {
            base.PropertyHasChanged(property);
            BusinessRules.CheckRules();
            BusinessRules.CheckObjectRules();
        }

        protected override void OnChildChanged(ChildChangedEventArgs e)
        {
            base.OnChildChanged(e);
            BusinessRules.CheckRules();
        }

I invoke the business rule stream but you can sum the values and they will be displayed when changed.

Regards,

Kevin

**Reply by Chicagoan2016** [reply]
@kcabral817 
Kevin, 
We have another use case which involves children and grand children. I am going to follow your approach but I was wondering what's the use of 
   ```cs
       protected override void PropertyHasChanged(IPropertyInfo property)
        {
            base.PropertyHasChanged(property);
            BusinessRules.CheckRules();
            BusinessRules.CheckObjectRules();
        }
  ```
?
Kind Regards

**Comment by kcabral817** [comment]
There are times when a specific property changes that I want to invoke the rules.  This was mainly for WPF and my not be applicable to Blazor

**Reply by rockfordlhotka** [reply]
@Chicagoan2016 also keep in mind that you can use the `ChildChanged` event to detect changes in children within your object graph. That exists specifically to aid in implementing scenarios where a parent needs to run rules based on the fact that a child or grandchild or whatever has been changed.

**Reply by Chicagoan2016** [reply]
@rockfordlhotka I think we have to use 'Target' property in a business rule like this or could we implement it without 'Target'?
Regards
